Readme.md
# Beam - BEware, Another Mailer #

Beam is a mail-client for BeOS/Haiku.

It is being developed by Oliver Tappe and distributed under the GPL
(see GnuGeneralPublicLicense.txt).

Software required by / used within Beam:

- The Layout Library (liblayout.so), by Marco Nelissen. Beam requires this to run!

- PCRE, Perl-Compatible-Regular-Expression library, by Philip Hazel. Beam requires
  this to run!

- Parts from SantasGiftBag, by Brian Tietz, hacked up a bit for use in Beam.

- Regexx, a C++-Regular-Expression library, by Gustavo Niemeyer, 
  ported to BString and hacked up a bit for use in Beam.

- Libiconv, a library that does conversion between different
  characterset-encodings, by Bruno Haible.

- OSBF from crm114, a bayes-like filter that's good for SPAM-filtering,
  by Fidelis Assis and Bill Yerazunis.

- openssl (if available), by the OpenSSL-team, is used to provide transport 
  layer security for all network traffic.

If you have questions, suggestions, complaints or a nice set of icons, 
you can reach the author at: <beam@hirschkaefer.de>.

Have fun!


Changes.txt
change-history of Beam:

< 2008-03-27: added tag 'rel-1-1-2' >

< 2008-03-27: commit >

BmMailEditWin:
	*	fixed another bug reported by Hartmut Reh: the edit window's title would
		not get updated when editing the subject of a mail

< 2008-03-26: commit >

BmMailFactory:
	*	Insane amount of whitespace within one line could steer the wrapping
		code into an endless loop. The wrapping code is still a mess, but it
		will handle this case more gracefully, now.
	*	implemented a suggestion by Ingo: smart replies should check
		automatically if it makes sense to do a reply-to-all
  
< 2008-03-23: commit >

BmMailEditWin, BmPrefsIdentityView:
	*	fixed bad bug when editing a mail's headers that would write data
		into the wild, causing crashes (noticed by Hartmut Reh - thanks!)

BmMenuAlert, ListSelectionAlert, TextEntryAlert:
	*	added sanity check to SetShortcut() for all alert classes

< 2007-10-23: commit >

BmPeople:
	*	added two missing sanity checks that could lead to crashes with people
		files that had only empty 'email...'-attributes (again reported by Rainer, 
		thanks!).

< 2007-10-22: commit >

BmPeople:
	*	fixed problem regarding people files that have email2 set but an empty 
		'email' attribute and their sorting into groups (reported by Rainer).

< 2007-10-19: commit >

BmMailEditWin:
	*	fixed incorrect handling of multiple mail-addresses in 
		AddAddressToTextControl() as reported by Rainer Riedl - thanks Rainer!

< 2007-10-02: commit >

MailConverter:
	*	put this back into a working state (triggered by Steffen Friedle asking
		me about such a tool)

< 2007-09-03: moved tag 'rel-1-1-1' >

< 2007-09-03: commit >

BmMailEditWin:
	*	added a warning before a mail without any subject is being sent
		[suggested by Stippi]
	*	turned a couple of logged errors into proper alerts, as they deserve
		that kind of attention (mail fields missing values)

< 2007-09-02: added tag 'rel-1-1-1' >

< 2007-09-02: commit >

BmBodyPartList, BmEncoding:
	*	fixed a really bad bug that caused problems with mails that contained 
		\xA0 as part of UTF-8 sequences. BmMailtextCleaner took them for 
		shift-space and replaced them by real spaces, which in turn invalidated 
		the UTF-8. [Bug spotted by Rainer a long time ago]

< 2007-08-23: commit >

BmSpamFilter:
	*	cleaned up (i.e. removed) several options for the SPAM-filter, as they
		should all be switched on at all times anyway
	*	added one more option that allows the user to select whether or not the
		SPAM-filter should stop processing the filter chain for mails considered
		SPAM
	*	used the free space to rearrange the options more pleasingly (for my
		liking, at least)

< 2007-08-22: commit >

BmSieveFilter:
	*	fixed bug causing infinite flooding with change messages for ListID 
		textcontrol, causing havoc and sometimes even crashes

< 2007-08-21: commit >

BmMail, BmMailFactory, BmMailHeader:
	*	Fixed problematic copying of mail-header when copying mails 
		("Edit As New..."), which could lead to problems when the mail is being 
		sent to a mailing list (that would be fooled into thinking the mail 
		looped). Now we rebuild the mail header based only on the basic fields
		(Bcc, Cc, From, Reply-To, Subject, To).

< 2007-05-09: commit >

BmMailHeader:
	*	minor improvement with respect to handling of List-Post field when trying
		to determine the list's address.

< 2007-02-09: commit >

BmMailHeader:
	*	when replying to a list, we now pay attention to the reply-to field, as
		this is required when a list redirects replies to another list

< 2006-11-18: commit >

BmGuiUtil, BmMailRefView, BmPrefs, BmPrefsGeneralView, BmPrefsShortcutsView
BmToolbarButton:
	*	added support for changing the shortcuts on-the-fly
	*	added two missing shortcuts (I'm pretty sure more are missing)
	*	refactored BmToolbarManager into something more reusable: BmViewManager

BmMainWindow:
	*	fixed typo in menu-entry

< 2006-11-12: commit >

BmIdentity:
	*	replaced very unreliable way of matching against aliases by a much more
		robust way. The basic idea was insane, as mails with mail-addresses that
		could not be used as regex would cause Beam to fail downloading this and
		any other mail from that specific account, doh!

BmImap:
	*	minor bugfix related to case-insensitivity of server responses

< 2006-11-10: commit >

BmImap, BmPopper, BmSmtp:
	-	fixed bug with SMTP not working at all when the user requests no 
		authentication at all. Thanks to Thom Holwerda for reporting this.
	-	adjusted POP3 and IMAP along the same lines, i.e. to work without 
		authentication (although it is much more unlikely that this will 
		ever be used).

< 2006-10-15: commit >

BmMail:
	-	added helper method ReconstructRawText() which updates the raw mailtext
		from all other values

BmMailFilter.cpp, BmSieveFilter:
	-	added support for setting the List-Id from within a matching filter. This
		is very handy in order to force mails to have come from specific lists 
		(which sometimes isn't the case if people do multiposting and the like).

BmPrefs:
	-	updated case of logging-states to our new convention

BmPrefsLoggingView:
	-	first batch of typing corrections

< 2006-09-08: commit >

BmSieveFilter:
	-	fixed bug with LearnAsSpam/Tofu-control not being usable at all (reported
		by Rainer)

< 2006-05-27: commit >

BetterScrollView:
	-	layout fix for Zeta-1.2

< 2006-05-22: added tag 'rel-1-1' >

< 2006-05-22: commit >

BmDataModel:
	-	TellJobIsDone() now explicitly sets the
		job-state of a non-completed job to stopped.
		This fixes problems with jobs being still 
		marked as running when in fact they were
		stopped because of an error. This in turn
		fixes the problem reported by Rainer: 
		Filtering a non-existing mail caused the
		filter-view to appear but never to finish
		(it was waiting for the mail to come out
		of the running state)

BmMailFilter:
	-	only ever execute filters on mails that
		have properly been initialized (other mails
		are ignored)

< 2006-05-21: commit >

BeamApp:

src-bmBase/md5c.c
/* MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm */

/* Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.
 */

/*
 * [zooey]: all this is just temporary and will be replaced by cryptlib
 *          in a later release (when Beam will support secure transport).
 */


#include <cstdio>
#include <cstring>

#define BM_BUILDING_BMBASE 1

#include "md5global.h"
#include "md5.h"

/* Constants for MD5Transform routine.
 */
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static void MD5Transform PROTO_LIST ((UINT4 [4], unsigned char [64]));
static void Encode PROTO_LIST
  ((unsigned char *, UINT4 *, unsigned int));
static void Decode PROTO_LIST
  ((UINT4 *, unsigned char *, unsigned int));
static void MD5_memcpy PROTO_LIST ((POINTER, POINTER, unsigned int));
static void MD5_memset PROTO_LIST ((POINTER, int, unsigned int));

static unsigned char PADDING[64] = {
  0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/* F, G, H and I are basic MD5 functions.
 */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits.
 */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
Rotation is separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

/* MD5 initialization. Begins an MD5 operation, writing a new context.
 */
void MD5Init (MD5_CTX *context)
                                        /* context */
{
  context->count[0] = context->count[1] = 0;
  /* Load magic initialization constants.
*/
  context->state[0] = 0x67452301;
  context->state[1] = 0xefcdab89;
  context->state[2] = 0x98badcfe;
  context->state[3] = 0x10325476;
}

/* MD5 block update operation. Continues an MD5 message-digest
  operation, processing another message block, and updating the
  context.
 */
void MD5Update (MD5_CTX *context, unsigned char *input, unsigned int inputLen)
                                        /* context */
                                /* input block */
                     /* length of input block */
{
  unsigned int i, index, partLen;

  /* Compute number of bytes mod 64 */
  index = (unsigned int)((context->count[0] >> 3) & 0x3F);

  /* Update number of bits */
  if ((context->count[0] += ((UINT4)inputLen << 3))
   < ((UINT4)inputLen << 3))
 context->count[1]++;
  context->count[1] += ((UINT4)inputLen >> 29);

  partLen = 64 - index;

  /* Transform as many times as possible.
*/
  if (inputLen >= partLen) {
 MD5_memcpy
   ((POINTER)&context->buffer[index], (POINTER)input, partLen);
 MD5Transform (context->state, context->buffer);

 for (i = partLen; i + 63 < inputLen; i += 64)
   MD5Transform (context->state, &input[i]);

 index = 0;
  }
  else
 i = 0;

  /* Buffer remaining input */
  MD5_memcpy
 ((POINTER)&context->buffer[index], (POINTER)&input[i],
  inputLen-i);
}

/* MD5 finalization. Ends an MD5 message-digest operation, writing the
  the message digest and zeroizing the context.
 */
void MD5Final (unsigned char digest[16], MD5_CTX *context)
                         /* message digest */
                                       /* context */
{
  unsigned char bits[8];
  unsigned int index, padLen;

  /* Save number of bits */
  Encode (bits, context->count, 8);

  /* Pad out to 56 mod 64.
*/
  index = (unsigned int)((context->count[0] >> 3) & 0x3f);
  padLen = (index < 56) ? (56 - index) : (120 - index);
  MD5Update (context, PADDING, padLen);

  /* Append length (before padding) */
  MD5Update (context, bits, 8);
  /* Store state in digest */
  Encode (digest, context->state, 16);

  /* Zeroize sensitive information.
*/
  MD5_memset ((POINTER)context, 0, sizeof (*context));
}

/* MD5 basic transformation. Transforms state based on block.
 */
static void MD5Transform (UINT4 state[4], unsigned char block[64])
{
  UINT4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];

  Decode (x, block, 64);

  /* Round 1 */
  FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
  FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
  FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
  FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
  FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
  FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
  FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
  FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
  FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
  FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
  FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
  FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
  FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
  FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
  FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
  FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

 /* Round 2 */
  GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
  GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
  GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
  GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
  GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
  GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
  GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
  GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
  GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
  GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
  GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
  GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
  GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
  GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
  GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
  GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

  /* Round 3 */
  HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
  HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
  HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
  HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
  HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
  HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
  HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
  HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
  HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
  HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
  HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
  HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
  HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
  HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
  HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
  HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

  /* Round 4 */
  II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
  II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
  II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
  II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
  II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
  II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
  II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
  II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
  II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
  II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
  II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
  II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
  II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
  II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
  II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
  II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;

  /* Zeroize sensitive information.
*/
  MD5_memset ((POINTER)x, 0, sizeof (x));
}

/* Encodes input (UINT4) into output (unsigned char). Assumes len is
  a multiple of 4.
 */
static void Encode (unsigned char *output, UINT4 *input, unsigned int len)
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4) {
 output[j] = (unsigned char)(input[i] & 0xff);
 output[j+1] = (unsigned char)((input[i] >> 8) & 0xff);
 output[j+2] = (unsigned char)((input[i] >> 16) & 0xff);
 output[j+3] = (unsigned char)((input[i] >> 24) & 0xff);
  }
}

/* Decodes input (unsigned char) into output (UINT4). Assumes len is
  a multiple of 4.
 */
static void Decode (UINT4 *output, unsigned char *input, unsigned int len)
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4)
 output[i] = ((UINT4)input[j]) | (((UINT4)input[j+1]) << 8) |
   (((UINT4)input[j+2]) << 16) | (((UINT4)input[j+3]) << 24);
}

/* Note: Replace "for loop" with standard memcpy if possible.
 */

static void MD5_memcpy (POINTER output, POINTER input, unsigned int len)
{
  unsigned int i;

  for (i = 0; i < len; i++)
 output[i] = input[i];
}

/* Note: Replace "for loop" with standard memset if possible.
 */
static void MD5_memset (POINTER output, int value, unsigned int len)
{
  unsigned int i;

  for (i = 0; i < len; i++)
 ((char *)output)[i] = (char)value;
}

/***********************************************************
 * MD5Digest (unsigned char *s)
 *    - added to this file by [zooey]
 *    - original code from Atsushi Takamatsu, only slightly modified
 ***********************************************************/

void MD5Digest (unsigned char *s, char* out) {
	int i;
	unsigned char digest[17];
	
	MD5Sum(s, (char*)digest);

  	for (i = 0;  i < 16;  i++) {
  		unsigned char c = digest[i];
		*out++ = (c > 0x9F ? 'a'-10 : '0')+(c>>4);
		*out++ = ((c&0x0F) > 9 ? 'a'-10 : '0')+(c&0x0F);
  	}
  	*out = '\0';
}

/***********************************************************
 * MD5Sum ()
 *    - added to this file by [zooey]
 *    - hashes the given string
 ***********************************************************/

void MD5Sum (unsigned char *s, char* out) {
	MD5_CTX context;
	
	MD5Init(&context);
	MD5Update(&context, s, strlen((char*)s));
	MD5Final((unsigned char*)out, &context);
	out[16] = '\0';
}



/***********************************************************
 * MD5_HMAC (text, key, digest)
 *    - added to this file by [zooey]
 *    - original code taken from RFC2104
 ***********************************************************/

void MD5_HMAC( unsigned char* text, int text_len, 
					unsigned char* key, int key_len, unsigned char* out)
{ 
	MD5_CTX context; 
	unsigned char k_ipad[65];    /* inner padding - 
	                              * key XORd with ipad 
	                              */ 
	unsigned char k_opad[65];    /* outer padding - 
	                              * key XORd with opad 
	                              */ 
	unsigned char tk[16]; 
	unsigned char digest[16]; 
	int i; 
	/* if key is longer than 64 bytes reset it to key=MD5(key) */ 
	if (key_len > 64) { 
	
		MD5_CTX      tctx; 
		
		MD5Init(&tctx); 
		MD5Update(&tctx, key, key_len); 
		MD5Final(tk, &tctx); 
		
		key = tk; 
		key_len = 16; 
	} 
	
	/* 
	 * the HMAC_MD5 transform looks like: 
	 * 
	 * MD5(K XOR opad, MD5(K XOR ipad, text)) 
	 * 
	 * where K is an n byte key 
	 * ipad is the byte 0x36 repeated 64 times 
	
	 * opad is the byte 0x5c repeated 64 times 
	 * and text is the data being protected 
	 */ 
	
	/* start out by storing key in pads */ 
	memset( k_ipad, 0, sizeof k_ipad); 
	memset( k_opad, 0, sizeof k_opad); 
	memcpy( k_ipad, key, key_len);
	memcpy( k_opad, key, key_len);
	
	/* XOR key with ipad and opad values */ 
	for (i=0; i<64; i++) { 
		k_ipad[i] ^= 0x36; 
		k_opad[i] ^= 0x5c; 
	} 
	/* 
	 * perform inner MD5 
	 */ 
	MD5Init(&context);                   /* init context for 1st 
	                                      * pass */ 
	MD5Update(&context, k_ipad, 64);     /* start with inner pad */ 
	MD5Update(&context, text, text_len); /* then text of datagram */ 
	MD5Final(digest, &context);          /* finish up 1st pass */ 
	/* 
	 * perform outer MD5 
	 */ 
	MD5Init(&context);                   /* init context for 2nd 
	                                      * pass */ 
	MD5Update(&context, k_opad, 64);     /* start with outer pad */ 
	MD5Update(&context, digest, 16);     /* then results of 1st 
	                                      * hash */ 
	MD5Final(digest, &context);          /* finish up 2nd pass */ 

  	for (i = 0;  i < 16;  i++) {
  		unsigned char c = digest[i];
		*out++ = (c > 0x9F ? 'a'-10 : '0')+(c>>4);
		*out++ = ((c&0x0F) > 9 ? 'a'-10 : '0')+(c&0x0F);
  	}
  	*out = '\0';
} 

